File: DFS/886. Possible Bipartition.py
# DFS
class Solution:
    def possibleBipartition(self, N: int, dislikes: List[List[int]]) -> bool:
        mp = collections.defaultdict(list)
        for i, j in dislikes:
            mp[i].append(j)
            mp[j].append(i)
        dp = [0 for _ in range(N+1)]
        def dfs(point, color):
            if dp[point]: 
                return dp[point] == color
            dp[point] = color
            return all( dfs(nxt, -color) for nxt in mp[point])
            
        return all(dfs(i,1) for i in range(1, N+1) if dp[i] == 0)


# BFS

class Solution:
    def possibleBipartition(self, N: int, dislikes: List[List[int]]) -> bool:
        mp = collections.defaultdict(list)
        for i, j in dislikes:
            mp[i].append(j)
            mp[j].append(i)
        
        dp = [0 for _ in range(N+1)]
        for i in range(1, N+1):
            q = collections.deque()
            if dp[i] == 0:
                q.append(i)
            color = 1
            while q:
                size = len(q)
                for _ in range(size):
                    top = q.popleft()
                    dp[top] = color
                    for nxt in mp[top]:
                        if dp[nxt] and dp[nxt] != -color:
                            return False
                        elif dp[nxt] == 0:
                            q.append(nxt)
                color = -color
        
        return True

File: DFS/886. Possible Bipartition.cpp
//DFS
class Solution {
public:
    bool possibleBipartition(int N, vector<vector<int>>& dislikes) {
        if(dislikes.empty())
            return true;
        unordered_map<int,vector<int>>mp;
        for(auto d: dislikes){
            mp[d[0]].push_back(d[1]);
            mp[d[1]].push_back(d[0]);
        }
        vector<int>dp(N+1,-1);
        for(int i = 1; i<=N; ++i)
            if(dp[i] == -1 && !dfs(mp, dp, 1, i))
                return false;
        return true;
    }
    
    bool dfs(unordered_map<int,vector<int>>&mp, vector<int>&dp, int color, int point){
        if(dp[point]!=-1)
            return dp[point] == color;
        dp[point] = color;
        for(auto next: mp[point])
            if(!dfs(mp, dp, color^1, next))
                return false;
        return true;
    }   
};


//BFS
// Time complexity: O(|V| + |E|)
// Space complxity: O(|V|)
class Solution {
public:
    bool possibleBipartition(int N, vector<vector<int>>& dislikes) {
        if(dislikes.empty())
            return true;
        unordered_map<int,vector<int>>mp;
        for(auto d: dislikes){
            mp[d[0]].push_back(d[1]);
            mp[d[1]].push_back(d[0]);
        }
        vector<int>dp(N+1,0);
        for(int i = 1; i<=N; ++i){
            queue<int>q;
            if(dp[i]==0)
                q.push(i);
            
            int color = 1;
            
            while(!q.empty()){
                int size = q.size();
                for(int j = 0; j<size; ++j){
                    int top = q.front(); q.pop();
                    dp[top] = color;
                    for(auto nxt: mp[top]){
                        if(dp[nxt]!= 0){
                            if(dp[nxt] != -color)
                                return false;
                        }else
                            q.push(nxt);
                    }
                }
                color *= -1;
            }
        }   
        return true;
    } 
};